### ISSUE TITLE
Add import functionality for IBM Watson Assistant (ex Conversation, ex Dialog)

### ISSUE THREAD

**Issue by muricaru** [RasaHQ/rasa_nlu#1154]
[IBM Watson Assistant] (https://www.ibm.com/watson/ai-assistant/)  is a popular solution to build ChatBots. 
We do have import from multiple systems - DialogFlow, Wit and Luis but we are missing an import procedure for Watson Assistant.
Please add the functionality to Rasa NLU 

@alexweidauer

**Comment by amn41** [comment]
is it possible to export your data from IBM Watson?  I think last time I checked it wasn't possible. If it is this would make a great contribution.

**Comment by muricaru** [comment]
Yes, I can provide you with a script capable to download programatically the workspace (this is the name IBM uses for the model). It will include intents, entities and the flow. On top of that will contain any hand made adjustments to the model. I can share some generic ones and some more advance in private as samples for the import procedures.

**Comment by ozlatior** [comment]
Hi,
The download script uses an API. It's nodejs only for now, but we can provide a python version as well if it makes more sense to you.
I just pushed it here together with a small sample: https://github.com/ozlatior/watson-import
There is also a sample workspace I've set up (it's actually a sample from IBM/Watson), the credentials are already in the config.

**Comment by akelad** [comment]
@ozlatior @muricaru any chance you are still interested in implementing this?

**Comment by ralphg6** [comment]
@akelad I am interested in implementing ...

I actually implemented it in Node.js, but in a slightly different way. I saw that the design approach is to code a loader that parses to TrainingData ... I'm moving to this architecture ...

But I've been thinking ... Who think that create a command in CLI to perform the migration already to the RASA NLU format definitively ...

Because what I realized was that the main use of this model is for proof of concept where we continue to use the source platform. But for cases where change is definite a migration actually makes more sense.

**Comment by erohmensing** [comment]
> I saw that the design approach is to code a loader that parses to TrainingData ... I'm moving to this architecture ...

Awesome! 

> . Who think that create a command in CLI to perform the migration already to the RASA NLU format definitively ...

Are you saying that in addition to being able to load Watson data directly, you'd like a CLI command to convert Watson to Rasa, similar to how we have one that converts json to md?

**Comment by ralphg6** [comment]
> Are you saying that in addition to being able to load Watson data directly, you'd like a CLI command to convert Watson to Rasa, similar to how we have one that converts json to md?

Exactly! What CLI command is this?

I thought of something like:
`rasa migrate --input-format watson --target-data-folder <data_folder>`

Using Loaders seems pretty simple!

**Comment by erohmensing** [comment]
So first thing's first, we would want to be able to load and use watson data directly as well, since we already support that for the other migrations. 


However I think this is an interesting idea for a separate enhancement -- @amn41 what do you think? Something like 
```
rasa data migrate
```
that would use the current guess_format to figure out the current format of the data out of the supported formats and then call a migration script to change it to rasa format?

**Comment by ralphg6** [comment]
I found the:
`rasa data convert`
Apparently I can already achieve this result!

**Comment by ralphg6** [comment]
Now on this issue, I already built the Watson Loader. In branch [migrate_watson](https://github.com/ralphg6/rasa/tree/migrate_watson).

However, as the process of Tokenization and Entity Extraction is very closely coupled with examples of intent, and in Watson it is not usual to mention entities in the examples, including entities not linked to examples.
Thus by converting Watson data to TrainingData format, the entities are not mentioned in the intent examples, and what happened was that the training did not recognize the entities.

In branch [found_others_entities](https://github.com/ralphg6/rasa/tree/found_others_entities) i had the other entities included in the dataset.

**Comment by ralphg6** [comment]
@erohmensing or @amn41 help me with this?

**Comment by erohmensing** [comment]
Hey @ralphg6 super cool that you're already working on this! I'll take a look at your branch -- can you clarify what you mean by

> in Watson it is not usual to mention entities in the examples, including entities not linked to examples.


And is it possible for you to share an excerpt of Watson data so that we can see what it looks like and what exactly you mean?

**Comment by ralphg6** [comment]
[skill-Teste-com-Rasa.json.txt](https://github.com/RasaHQ/rasa/files/3467529/skill-Teste-com-Rasa.json.txt)

**Comment by ralphg6** [comment]
In Watson, the conventional is to register the entities and intentions separately.
It is not common to create examples of intentions using entities, but clear text.
That is, there is no relationship between entities and intentions!

**Comment by ralphg6** [comment]
I will attach from file generated since JSON so you can see how it looks. Thus entities need to be converted as synonyms or LookupTables. But since they are not intentional, they are not in the training dataset.

**Comment by ralphg6** [comment]
I am uploading the files again what I had uploaded, I had modified it for a test.
[skill-Teste-com-Rasa.json.txt](https://github.com/RasaHQ/rasa/files/3467604/skill-Teste-com-Rasa.json.txt)
[generated.nlu.md.txt](https://github.com/RasaHQ/rasa/files/3467607/generated.nlu.md.txt)

**Comment by ralphg6** [comment]
I made three branches:

[migrate_watson](https://github.com/ralphg6/rasa/tree/migrate_watson): This has Watson migration implementation

[found_others_entities](https://github.com/ralphg6/rasa/tree/found_others_entities): This includes the synonym entities in the training dataset.

[watson_and_entities](https://github.com/ralphg6/rasa/tree/watson_and_entities): It's the union of the previous two, I'm using this.

**Comment by akelad** [comment]
Hi @ralphg6, taking over from ella - so if I understand correctly with Watson you simply provide a list of entities separately (not in sentences) to your intent examples? 
That causes a small problem, because while we could use lookup tables, you would still need to add some sort of annotations to your training data in Rasa since it won't recognise the entities otherwise. 
One option would be to hard tag the entities from the list when converting to the rasa format - but let me discuss with the team whether that's a good idea first

**Comment by akelad** [comment]
so I think we can do hard tagging of the entities. @ralphg6 any progress so far?

**Comment by ralphg6** [comment]
Hi @akelad, i'm sorry for disappearing! I am back to win this barrier!

For adjust the compreension. We already converted Watson data for Rasa NLU data fomart. But Rasa's entity extractors need entity samples in intent sentences, because this extractors recognizes mainly entities per position in sentence. And in Watson the entities are registred alone as in a dictionary.

Then we need code a entity extractor able to train with this dictionary and recognize entity with only this. 

Obs: In Watson, the entities can be recognize per exact simillarity or fuzing matching. Entities with fuzing matching disabled should not be recognized without exact simillarity.

**Comment by ralphg6** [comment]
Someone has some ideia of algoritm able to train with this dictionary only?

I was suggested the cosine similarity with tf-idf, but yet i not have knowledge for to decise.

**Comment by akelad** [comment]
i think it's best to do hard tagging of entities for this dictionary. Or put it in a lookup table, but that still requires some example sentences

**Comment by ralphg6** [comment]
But I have to import a lot of data in this format. And the effort to create sample sentences is very large due to the amount of data I have.

```md
## synonym:car  
- vehicle
- automobile

## synonym:balance
- amount of money
- money remaining in account
```

This is sample of all data i have. I not able of create a sample sentence for each entity synonym. 

And i know is possible extract this entities with only this data, because in the Watson works.

Beyond if the user to utter 'My veihcle is broken' the entity 'car' should be recognized.

**Comment by akelad** [comment]
i mean this isn't quite how rasa NLU works though, so you will need to create some sample sentences for entities. Not for each one though, as you say.

**Comment by ralphg6** [comment]
I understand that Rasa NLU not works in that way. 
I am suggesting a new feature.
And i am studying for to contribute with this. 

@akelad did you believe Rasa NLU can not has this feature?

**Comment by akelad** [comment]
you mean the feature of fuzzy matching or which one?

**Comment by ralphg6** [comment]
if the user to utter 'My veihcle* is broken' _with typo error_ the entity 'car' should be recognized, this is fuzzy matching mode on.

**Comment by akelad** [comment]
yeah, i mean we've had contributions from the community on this before, but we decided against it. That was because after running some experiments, we found it to be quite slow. You can always use a custom component for this though

**Comment by ralphg6** [comment]
Understand! But the entities data in Watson Assistant are in this format, if the verdict is: do not implement. I can't see a solution for import from Watson.

**Comment by tabergma** [comment]
@ralphg6 Just a small note: We are working on improving our NER model right now and the results look very promising. Thus, our future NER model should be better with respect to typos 😄

**Comment by akelad** [comment]
@ralphg6 i mean i think that you can still implement this import, like i said you can hard tag the entities that you have defined. Sure, it may not have the typos directly in there, but it's certainly better than nothing

**Comment by ralphg6** [comment]
@akelad to improve my understanding, what do you mean by 'tagging entities'?
I question because I understand that in the import process Rasa NLU should work without manual intervention. Important fact, because in cases equals as ours, that has lot of data, the manual intervention invalid the import, for is very expensive.

**Comment by akelad** [comment]
i meant with a hard tagging script. So taking the list of entities from watson, and running a script to hard tag all the entities in the sentences that are listed there
